main.cpp

```diff
@@ -402,3 +403,157 @@ int main(int argc, char** argv) {
+        printf("Failed to restore cert!\n");
+        cert_strategy.patch_action = P_DO_NOTHING;
+        while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_BACKUP_FAIL); PresentMenu(); }
+        return -1;
+    }
+ RenderMenuLoading(task_percent(task++), "Verifying final patches... (wave.rpx)");
+    PresentMenu();
+
+    {
+        std::ifstream is(wave_path, std::ios::binary);
+        auto hash = rpx_hash(is);
+        if (hash.patch != RPX_PATCH_STATE_STOCK) {
+            printf("Failed to commit patches - stock wave in place\n");
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_RESTORE_FAIL); PresentMenu(); }
+            return -1;
+        } else if (hash.id == CURRENT_PRETENDO_WAVE) {
+            printf("Failed to commit patches - Pretendo wave in place\n");
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_RESTORE_FAIL); PresentMenu(); }
+            return -1;
+        }
+    }
+    RenderMenuLoading(task_percent(task++), "Verifying final patches... (nn_olv.rpl)");
+    PresentMenu();
+
+    {
+        std::ifstream is(nn_olv_path, std::ios::binary);
+        auto hash = rpx_hash(is);
+        if (hash.id == CURRENT_PRETENDO_NN_OLV) {
+            //we've still have the PN OLV
+            printf("Failed to commit patches - nn_olv corrupt!\n");
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_RESTORE_FAIL); PresentMenu(); }
+            return -1;
+        }
+    }
+   RenderMenuLoading(task_percent(task++), "Verifying final patches... (NSSL cert)");
+    PresentMenu();
+
+    {
+        std::ifstream is(cert_path, std::ios::binary);
+        auto hash = cert_hash(is);
+        if (hash.id != CERT_ID_THWATE_PREMIUM_SERVER) {
+            //we've already modified wave, so this is always dangerous
+            printf("Failed to commit patches - nssl cert corrupt!\n");
+            //oh shit, this could brick - restore backup
+            bret = fast_copy_file(cert_bak_path, cert_path);
+            if (!bret) {
+                printf("Reverting nssl patches failed!\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_CERT_BRICK); PresentMenu(); }
+                return -1;
+            }
+
+            std::ifstream is2(cert_path, std::ios::binary);
+            auto hash2 = cert_hash(is2);
+            if (hash2.patch != CERT_PATCH_STATE_STOCK) {
+                printf("Reverting nssl patches failed!\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_CERT_BRICK); PresentMenu(); }
+                return -1;
+            } else if (hash2.id != CURRENT_SACRIFICIAL_CERT) {
+                printf("Reverting NSSL cert resulted in the wrong cert?\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_DANGEROUS); PresentMenu(); }
+                return -1;
+            }
+
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_DANGEROUS); PresentMenu(); }
```

Comment:
This may not be dangerous at this point since we've restored wave? Not 100%, will have to think through what the different possibilities are.